src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs

```diff
@@ -569,6 +569,21 @@ public BoundNode VisitStatement(BoundNode node)
             return VisitSideEffect(node);
         }
 
+        public ImmutableArray<BoundStatement> VisitSideEffects(ImmutableArray<BoundStatement> statements)
+        {
+#if DEBUG
+            var result = ArrayBuilder<BoundStatement>.GetInstance(statements.Length);
+            foreach (BoundStatement statement in statements)
+            {
+                result.Add((BoundStatement)VisitSideEffect(statement));
+            }
+
+            return result.ToImmutableAndFree();
+#else
+            return this.VisitList(statements);
```

Comment:
>return this.VisitList(statements); [](http://example.com/codeflow?start=12&length=34)

What is the benefit in doing something different for DEBUG?

Reply:
We're just trying to bypass stack assertion in VisitStatement. This is structurally equivalent to EmitSideEffects.

Reply:
> We're just trying to bypass stack assertion in VisitStatement. This is structurally equivalent to EmitSideEffects.

I think we should find a way to avoid the code duplication and  possibility to diverge in the future. In both places, and in other similar places if they were introduced in this PR.

Reply:
> How would you imagine it could be done? 

I suppose the visitor could have state that describes what stack depth to assert. That state could be debug only. Perhaps there are other ways.

Reply:
That doesn't remove the duplication though, it just means more code in each class to bypass a Debug.Assert. If that's fine I can try it for both CodeGenerator and Optimizer types but I believe the current revision is sufficiently clear.

Reply:
>  I believe the current revision is sufficiently clear.

Clarity is not my primary concern in this case, code maintenance is.

Reply:
Pushed a commit to address two issues:
- Adjusting the assertion instead of working around it
- Covering nested statements e.g. StatementList